Web diagnostics

dominikl opened the pull request:
# What this PR does

`./omero admin diagnostics` also checks status and logs of OMERO.web, but as OMERO.web could be run independantly, it falsely reports an error in that case.
This PR removes the 'web' part from 'admin diagnostics' and puts it into the web module as 'web diagnostics'

# Testing this PR

- Run `./omero admin diagnostics` and check that you don't see any web related output.
- Run `./omero web diagnostics` and check that you see all the web related output.

# Related reading

https://trello.com/c/XdIDdBNS/137-omero-admin-diagnostics-web-status-is-misleading

dominikl committed:
Move 'web diagnostics' part from admin to web

dominikl committed:
Move web log check diagnostics from admin to web

dominikl commented:
I did quite a lot copy/pasting (methods like `sz_str`, `item`) from admin.py to web.py. Guess there's a better way to do this, worth it?

sbesson commented on components/tools/OmeroPy/src/omero/plugins/web.py:
```diff
@@ -686,6 +694,80 @@ def set_environ(self, ice_config=None):
+
+        # OMERO.web diagnostics
+        self.ctx.out("")
+        from omero.plugins.web import WebControl
```
Since this is now a method in `Webcontrol`, this import is unecessary and `status(args)` can be used directly 2 lines below

sbesson reviewed (commented):
Agreed re  code duplication. Thinking of the ongoing micro-services, I could imagine some of this logic re-use more generically e.g. for an `bin/omero thumbnails-ms diagnostics`.

An immediate way to refactor would be to derive a new subclass of `BaseControl` with the helper methods in `cli.py` and make `AdminControl` and `WebControl`  inherit it.

joshmoore commented:
Something for a follow-on:

> An immediate way to refactor would be to derive a new subclass of BaseControl with the helper methods in cli.py and make AdminControl and WebControl inherit it.

`ctx` provides access to all `Control` objects. Perhaps instead we make an interface `DiagnosticsControl` and use multiple inheritance. Then `bin/omero diagnostics` could loop through all such controls.

dominikl committed:
Moved common methods to BaseControl

dominikl commented:
Moved the common methods to `BaseControl`. Is that better? By the way, I'm getting bit out of my python comfort zone here... ;-) Edit: ~~And it didn't work apparently.~~ Was just flake8

dominikl committed:
Just fixed flake8 issues

joshmoore committed:
Make `diagnostics` subcommands discoverable

WebControl and AdminControl now also subclass DiagnosticsControl
so that the top-level method `bin/omero diagnostics` can now find
each and print out a concatenated output. Further code from admin.py
(like the upgrade check) may should similarly be refactored out.

joshmoore commented:
Pushed 11413f02d380b94c590b02a3a7b441a6dbfc115f for discussion. To some extent, this _reintroduces_ the confusion of `bin/omero admin diagnostics`, so we'd need to decide if this "advanced usage" is worth it. Regardless, it does introduce a more formal "diagnostics API" which may be worthwhile.

cc: @jburel @sbesson

joshmoore committed:
Migrate collecting diagnostics to a plugin

joshmoore commented:
Opened https://github.com/dominikl/openmicroscopy/pull/9 *without* the `bin/omero diagnostics` command (which I'll add to github.com/joshmoore/omero-cli-diagnostics)

dominikl committed:
Merge pull request #9 from joshmoore/web_diagnostics

Web diagnostics

jburel commented:
@dominikl can we retest this PR?

dominikl commented:
Yes, I think so. Sorry should have listed it on standup.

manics commented on components/tools/OmeroPy/src/omero/cli.py:
```diff
@@ -944,6 +944,75 @@ def _order_and_range_ids(self, ids):
+        diagnostics = parser.add(
+            sub, self.diagnostics,
+            "Run a set of checks on the current, "
+            "preferably active ")
```
preferably active... `server`? `application`?
```
$ omero admin --help
...
  <subcommand>
    diagnostics       Run a set of checks on the current, preferably active
    start             Start icegridnode daemon and waits for required components to come up, i.e. status == 0
...
```

manics reviewed (changes requested):
Looks fine apart from the help text.

dominikl committed:
Fixed help text

manics commented:
Good to merge
